### PR TITLE
AST-281: Set reference-type

### PR DESCRIPTION
**Pull request by pierallard**

### CODE REVIEW

File: src/Command/CreateFamilyCommand.php [hunk 1]
```diff
@@ -40,14 +45,32 @@ protected function configure()
     {
         $this
             ->setDescription('Create an Asset Family')
-            ->addArgument('assetFamilyCode', InputArgument::REQUIRED, 'The asset family code to create')
+            ->addArgument('asset-family-code', InputArgument::REQUIRED, 'The asset family code to create')
+            ->addOption('reference-type', null, InputOption::VALUE_OPTIONAL,
+                sprintf(
+                    'Enable if reference is localizable or not. Allowed values: %s|%s|%s',
```
Comment (SamirBoulil) [inline src/Command/CreateFamilyCommand.php]:
```suggestion
                    'Enable if image reference is localizable or not. Allowed values: %s|%s|%s',
```

File: src/Command/CreateFamilyCommand.php [hunk 2]
```diff
@@ -40,14 +45,32 @@ protected function configure()
...
     }
 
     public function execute(InputInterface $input, OutputInterface $output)
     {
         $this->io = new SymfonyStyle($input, $output);
-        $this->assetFamilyCode = $input->getArgument('assetFamilyCode');
+        $this->assetFamilyCode = $input->getArgument('asset-family-code');
+        $referenceType = $input->getOption('reference-type');
+        if (!in_array($referenceType, [self::LOCALIZABLE, self::NON_LOCALIZABLE, self::BOTH])) {
+            throw new \InvalidArgumentException(sprintf(
```
Comment (SamirBoulil) [inline src/Command/CreateFamilyCommand.php:69]: There is a way of stating this with symfony so you don't need to make the check here.

File: src/Command/CreateFamilyCommand.php [hunk 3]
```diff
@@ -57,28 +80,44 @@ public function execute(InputInterface $input, OutputInterface $output)
...
             'code' => $this->assetFamilyCode,
             'labels' => ['en_US' => $this->assetFamilyCode], // TODO AST-239 Need to set at least 1 label, else the UI fail :/
         ]);
 
-        $this->createAttribute('reference', 'media_file', false, true, false);
-        $this->createAttribute('reference_localizable', 'media_file', true, true, false);
-        $this->createAttribute('variation_scopable', 'media_file', false, true, false);
-        $this->createAttribute('variation_localizable_scopable', 'media_file', true, true, false);
+        if ($referenceType === self::NON_LOCALIZABLE) {
+            $this->createAttribute('reference', 'media_file', false, true, false);
```
Comment (SamirBoulil) [inline src/Command/CreateFamilyCommand.php]: can extract this bit into a createNonLocalizableAttributes

File: src/Command/CreateFamilyCommand.php (hunk 1 above)
Comment (juliensnz) [inline src/Command/CreateFamilyCommand.php]: We could ask for more details to delphine

File: src/Command/CreateFamilyCommand.php [hunk 4]
```diff
@@ -57,28 +80,44 @@ public function execute(InputInterface $input, OutputInterface $output)
...
+            $this->createAttribute('variation_scopable', 'media_file', false, true, false);
+        };
+
+        if ($referenceType === self::LOCALIZABLE) {
+            $this->createAttribute('reference_localizable', 'media_file', true, true, false);
```
Comment (SamirBoulil) [inline src/Command/CreateFamilyCommand.php]: Can also extract it here

File: src/Command/CreateFamilyCommand.php [hunk 5]
```diff
@@ -57,28 +80,44 @@ public function execute(InputInterface $input, OutputInterface $output)
...
+            $this->createAttribute('variation_localizable_scopable', 'media_file', true, true, false);
+        }
+
+        if ($referenceType === self::BOTH) {
+            $this->createAttribute('reference', 'media_file', false, true, false);
```
Comment (SamirBoulil) [inline src/Command/CreateFamilyCommand.php]: Call two functions here

File: src/Command/CreateFamilyCommand.php [hunk 6]
```diff
@@ -40,14 +45,32 @@ protected function configure()
...
+                self::BOTH
+            )
         ;
     }
 
     public function execute(InputInterface $input, OutputInterface $output)
     {
         $this->io = new SymfonyStyle($input, $output);
-        $this->assetFamilyCode = $input->getArgument('assetFamilyCode');
+        $this->assetFamilyCode = $input->getArgument('asset-family-code');
```
Comment (juliensnz) [inline src/Command/CreateFamilyCommand.php:66]: maybe we could extract that in a dedicated method. I leave it as you prefer

File: src/Command/CreateFamilyCommand.php [hunk 7]
```diff
@@ -57,28 +80,44 @@ public function execute(InputInterface $input, OutputInterface $output)
...
+            $this->createAttribute('variation_localizable_scopable', 'media_file', true, true, false);
+        }
+
         $this->createAttribute('description', 'text', false, false, false);
         $this->createAttribute('categories', 'text', false, false, false);
         $this->createAttribute('tags', 'text', false, false, false);
         $this->createAttribute('end_of_use', 'text', false, false, false);
 
-        $this->io->writeln('Update attribute as main media...');
+        $attributeAsMainMedia = $referenceType === self::NON_LOCALIZABLE || $referenceType === self::BOTH ? 'reference' : 'reference_localizable';
```
Comment (juliensnz) [inline src/Command/CreateFamilyCommand.php]: Could we extract 'reference' and 'reference_localizable' in consts?

File: src/Command/MergeAssetsAndVariationsFilesCommand.php [hunk 8]
```diff
@@ -57,25 +62,47 @@ class MergeAssetsAndVariationsFilesCommand extends Command
...
-        $assetsFilePath = $input->getArgument('assetsFilePath');
-        $variationsFilePath = $input->getArgument('variationsFilePath');
-        $targetFilePath = $input->getArgument('targetFilePath');
+        $assetsFilePath = $input->getArgument('assets-file-path');
+        $variationsFilePath = $input->getArgument('variations-file-path');
+        $targetFilePath = $input->getArgument('target-file-path');
         $this->variationsFilePath = $variationsFilePath;
 
+        $this->referenceType = $input->getOption('reference-type');
+        if (!in_array($this->referenceType, [self::LOCALIZABLE, self::NON_LOCALIZABLE, self::BOTH])) {
```
Comment (SamirBoulil) [inline src/Command/MergeAssetsAndVariationsFilesCommand.php:97]: Same, there is a way in symfony to put valid argument values

File: src/Command/CreateFamilyCommand.php [hunk 9]
```diff
@@ -57,28 +80,44 @@ public function execute(InputInterface $input, OutputInterface $output)
...
+        $this->io->writeln(sprintf('Update "%s" attribute as main media...', $attributeAsMainMedia));
+
         $this->client->getAssetFamilyApi()->upsert($this->assetFamilyCode, [
             'code' => $this->assetFamilyCode,
-            'attribute_as_main_media' => 'reference',
+            'attribute_as_main_media' => $attributeAsMainMedia,
         ]);
 
-        $this->io->success(sprintf('Family "%s" created', $this->assetFamilyCode));
+        $this->io->success(sprintf('Family "%s" created!', $this->assetFamilyCode));
```
Comment (juliensnz) [inline src/Command/CreateFamilyCommand.php:120]: Yeah!

**Review by SamirBoulil — APPROVED** [review]
Make sure you try to make it work on a real environment.

File: src/Command/MergeAssetsAndVariationsFilesCommand.php [hunk 10]
```diff
@@ -230,11 +261,28 @@ private function mergeAssetAndVariations(array $oldAsset, array $variations): ar
...
             } else {
-                $structure[sprintf('%s-%s', self::REFERENCE_FILE_FIELD, $variation['channel'])] = $variation['reference_file'];
-                $structure[sprintf('%s-%s', self::VARIATION_FILE_FIELD, $variation['channel'])] = $variation['variation_file'];
+                if ($this->referenceType === self::NON_LOCALIZABLE || $this->referenceType === self::BOTH) {
+                    $structure[sprintf('%s-%s', self::REFERENCE_FILE_FIELD, $variation['channel'])] = $variation['reference_file'];
+                    $structure[sprintf('%s-%s', self::VARIATION_FILE_FIELD, $variation['channel'])] = $variation['variation_file'];
+                } else {
+                    throw new \RuntimeException(sprintf(
+                        "The merge script encountered an issue with \"%s\".
+                        \nThis line contains does not contains any value in the locale column, but this value is needed but it is needed for the asset family.",
```
Comment (juliensnz) [inline src/Command/MergeAssetsAndVariationsFilesCommand.php]: it contains or does not contains?

Comment (juliensnz) [reply src/Command/MergeAssetsAndVariationsFilesCommand.php]: and is it needed or but is it needed?

Comment (juliensnz) [reply src/Command/MergeAssetsAndVariationsFilesCommand.php]: I think you mixed your copy and paste

File: src/Command/MigrateCommand.php [hunk 11]
```diff
@@ -49,6 +63,16 @@ protected function configure()
             ->addArgument('asset-family-code', InputArgument::REQUIRED, 'The asset family code to migrate')
             ->addArgument('assets-csv-filename', InputArgument::OPTIONAL, 'The path to the Assets CSV file', self::ASSETS_CSV_FILENAME)
             ->addArgument('variations-csv-filename', InputArgument::OPTIONAL, 'The path to the Variations CSV file', self::VARIATIONS_CSV_FILENAME)
+            ->addOption('reference-type', null, InputOption::VALUE_OPTIONAL,
+                sprintf(
+                    'Enable if reference is localizable or not. Allowed values: %s|%s|%s|%s',
```
Comment (juliensnz) [inline src/Command/MigrateCommand.php]: could be nice to explain a bit more. Especially "auto"

File: src/Command/MigrateCommand.php [hunk 12]
```diff
@@ -89,4 +128,63 @@ private function executeCommand($name, $arguments)
...
+                    $foundNonLocalized = true;
+                }
+            }
+
+            if (!$foundLocalized && !$foundNonLocalized) {
+                $this->io->error('No assets found. This script can not guess the reference type.');
+
+                exit(1);
+            } else if ($foundNonLocalized && $foundLocalized) {
+                $this->io->writeln(sprintf("Localized and non localized assets found, set reference type to %s.", self::BOTH));
```
Comment (juliensnz) [inline src/Command/MigrateCommand.php]:
```suggestion
                $this->io->writeln(sprintf('Localized and non localized assets found, set reference type to "%s".', self::BOTH));
```

File: src/Command/MigrateCommand.php [hunk 13]
```diff
@@ -89,4 +128,63 @@ private function executeCommand($name, $arguments)
...
+                return self::BOTH;
+            } else if ($foundLocalized) {
+                $this->io->writeln(sprintf("Only localized assets found, set reference type to %s.", self::LOCALIZABLE));
```
Comment (juliensnz) [inline src/Command/MigrateCommand.php]: same here

File: src/Command/MigrateCommand.php [hunk 14]
```diff
@@ -89,4 +128,63 @@ private function executeCommand($name, $arguments)
...
+                return self::LOCALIZABLE;
+            } else {
+                $this->io->writeln(sprintf("Only non localized assets found, set reference type to %s.", self::NON_LOCALIZABLE));
```
Comment (juliensnz) [inline src/Command/MigrateCommand.php]: same here

**Review by juliensnz — APPROVED** [review]

File: src/Command/CreateFamilyCommand.php (hunk 2 above)
Comment (pierallard) [reply src/Command/CreateFamilyCommand.php:69]: Seen IRL, not found :(

**Pull request merged**